exonum/src/runtime/communication_channel.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::runtime::{
+    dispatcher::Dispatcher,
+    mailbox::{Action, AfterRequestCompleted, BlockchainMailbox},
+    CallInfo, ConfigChange, ExecutionContext, ExecutionError, InstanceId,
+};
+
+/// Marker struct for `Supervisor` entities.
+/// It is used as a type parameter for `CommunicationChannel` to provide
+/// `Supervisor`-related functionality.
+#[doc(hidden)]
+#[derive(Debug)]
+pub struct SupervisorAccess {}
+
+/// Communication channel is an proxy entity for performing service calls from
+/// other service instances.
+#[derive(Debug)]
+pub struct CommunicationChannel<'a, T = ()> {
+    pub(crate) mailbox: &'a BlockchainMailbox,
+    dispatcher: &'a Dispatcher,
+    phantom: std::marker::PhantomData<T>,
+}
+
+impl<'a, T> CommunicationChannel<'a, T> {
+    /// Creates a new communication channel.
+    pub(crate) fn new(mailbox: &'a BlockchainMailbox, dispatcher: &'a Dispatcher) -> Self {
+        Self {
+            mailbox,
+            dispatcher,
+            phantom: std::marker::PhantomData,
+        }
+    }
+
+    /// Calls the corresponding service method.
+    pub fn call(
+        &self,
+        context: &ExecutionContext,
+        call_info: &CallInfo,
```

Comment:
If this is to be used by services, and service name is considered the primary identifier of a service (and it is definitely easier to use during configuration of a service), why don't we pass service name instead of service id (in CallInfo)? If this method accepts id, then the callee has to either lookup it (how?) or, which is worse in terms of UX, the user will have to configure it with id of the service on which it depends (instead of the name, which the user already knows).

Reply:
This is to be used by runtimes (since it can provide supervisor extensions, it's dangerous to provide communication channel to the service directly). Runtimes are free to build a handy interface atop of `CommunicationChannel`.

Reply:
Ok, but the runtimes themselves also need the handy interface, at least, to translate names to ids (they can't do that on their own in case of cross-lang interop).

Reply:
Well, to be honest, I don't understand the difference between a numeric ID and a string name.

For example, you have a service A that wants to call service B.
Service A has 3 different instances, service B also has 3 different instances.
How instances of service A will know the name of instances of service B that they want to call?

I guess, it will be in the service config. And if you can provide a service instance name to the config, why can't you provide a numeric ID instead?

Reply:
You-the-user? Doesn't the framework assign the numeric ids?

Reply:
Yep, "you" == "the user".
And yep, numeric IDs are assigned by the framework.

Reply:
Then I see no reason to be user-administrator-hostile and require specifying a second identifier which the framework already knows; or user-service-developer-hostile and require the service to read it from _somewhere_. Administrators supply a name when they start service B. The name is the easiest way to reference B in the config for A.

Reply:
OK. Do I understand it correctly that you suggest to include a service name into the `CallInfo` structure which is a part of every transaction?
Or do you suggest to call the service through `communication_channel` in the different way than it'll be done through usual transaction?

Reply:
I don't suggest including in the transaction, but changing the signature of this method to use name instead of id.

Reply:
Then again, why is it not "hostile" to send a numeric ID in the transaction and it is "hostile" to use numeric ID here?

Reply:
Both are. For txs, IIRC, that is done for storage efficiency. For calls, which happen in a completely different context, I don't see a reason.

Reply:
Too bad you can't see the worse UX.

@aleksuss , would you please design the inter-service communication in core to have a reasonable way for a Runtime implementation to perform calls by service name?

Reply:
@dmitry-timofeev Ok. We'll discuss it.